### PR TITLE
Try: Use rule processing for remote payment methods

### PR DESCRIPTION
**Pull request by joshuatf**
This is a POC to use the existing rule processing in the remote inbox for fetching remote payment methods.

**Things to consider**

* Moving rules to its own separate class outside of remote inbox?  I'm in favor of this as I think these are highly reusable.
* Should this be onboarding payments (as it currently is)?  Should it be a generic payments endpoint?  What other endpoints might we need here?

**Todo**

* Cache time for fetched methods.
* Using API response on frontend.

### Detailed test instructions:

1. Add a filter to filter the data source:
```php
add_filter( 'woocommerce_admin_remote_payment_methods_data_sources', function() {
	return array(
		'http://yoursite.local/payment-methods.json',
	);
} );
```
2. In your `payment-methods.json` file, add some test data.
```json
[
    {
        "slug": "test-method",
        "locales": [
            {
                "locale": "en_US",
                "title": "Test Poll Sources",
                "content": "I'm a remote payment method"
            }
        ],
        "is_visible": [
            {
                "type": "option",
                "option_name": "woocommerce_currency",
                "value": "USD",
                "default": "USD",
                "operation": "="
            }
        ],
        "is_configured": [
            {
                "type": "option",
                "option_name": "my_payment_option",
                "value": "false",
                "default": "false",
                "operation": "!="
            }
        ]
    }
]
```
3. Attempt to fetch the methods via the new REST endpoint `/wp-json/wc-admin/onboarding/payments` (note you will need to delete the option `woocommerce_admin_remote_payment_methods_specs` if you update this file).
4. Try changing your site's options to pass or fail the rules (set to and from USD and provide any value for `my_payment_option` in the above example) to verify the correct boolean values are provided for `is_visible` and `is_configured`.

### CODE REVIEW

**Comment by becdetat** [comment]
I've only gone through the code, haven't tried running it, but it looks like an awesome feature. I like the idea of moving the rule processors out of RemoteInboxNotifications to make them reusable.

File: src/Features/RemotePaymentMethods/DataSourcePoller.php [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	/**
+	 * Reads the data sources for specs and persists those specs.
+	 *
+	 * @return bool Whether any specs were read.
+	 */
+	public static function read_specs_from_data_sources() {
+		$specs        = array();
+		$data_sources = apply_filters( 'woocommerce_admin_remote_payment_methods_data_sources', self::DATA_SOURCES );
```
Comment (timmyc) [inline src/Features/RemotePaymentMethods/DataSourcePoller.php:46]: 😍

**Comment by psealock** [comment]
Looking good @joshuatf, I'm trying to test this out. I ran out of time today, but thought I'd surface a question I have so far:

* Instead of adding a filter, I just changed the url in the code to this gist https://gist.githubusercontent.com/psealock/25f97019f3fc719c1e1bbffa4f348c19/raw/cf7eb75b162422314da81d82d17c28df8600a4f2/remote-payment-recs.json
* `wp-json/wc-admin/onboarding/payments` returns an empty array. I have my currency set to USD and my address in the US. 
* `wp option get woocommerce_admin_remote_payment_methods_specs` returns an empty array as well. 

Under these rules, I should be getting some payment methods returned, right?

**Comment by psealock** [comment]
Nice, a quick run through with a [gist I created](https://gist.githubusercontent.com/psealock/25f97019f3fc719c1e1bbffa4f348c19/raw/cf7eb75b162422314da81d82d17c28df8600a4f2/remote-payment-recs.json) is working well. I'll dive in deeper tomorrow

**Comment by timmyc** [comment]
One more little bit of food for thought... could we possibly use this same feed, and approach, to power the business extensions part of the onboarding flow? I suppose we could have two different feeds that both leverage the same rule processing, but just wanted to toss the idea out there :)

**Comment by becdetat** [comment]
@timmyc 
> could we possibly use this same feed, and approach, to power the business extensions part of the onboarding flow?

I would like to know your reasoning, but I personally would prefer having one endpoint per feature, rather than one feed to rule them all - single responsibility principle. The same _approach_ would be great to see though - I don't think there would be any need for rule processing for the onboarding flow but the rest of the approach could be reused.

**Review by psealock — APPROVED** [review]
This is looking great and working well in my testing. I changed the specs as well as criteria like currency to see the result of the GET request update accordingly. Nice one @joshuatf !

File: src/Features/RemotePaymentMethods/Init.php [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	 */
+	public static function get_specs() {
+		$specs = get_transient( self::SPECS_TRANSIENT_NAME );
+
+		// Fetch specs if they don't yet exist.
+		if ( false === $specs || ! is_array( $specs ) || 0 === count( $specs ) ) {
+			// We are running too early, need to poll data sources first.
+			$specs = DataSourcePoller::read_specs_from_data_sources();
+			$specs = self::localize( $specs );
+			set_transient( self::SPECS_TRANSIENT_NAME, $specs, 7 * DAY_IN_SECONDS );
```
Comment (psealock) [inline src/Features/RemotePaymentMethods/Init.php:59]: Would it be possible to set this transient to something super small for dev environments? Maybe we could go off `WP_DEBUG`?

Comment (joshuatf) [reply src/Features/RemotePaymentMethods/Init.php:59]: Good question.  As part of the testing in #6867 I added a remote payments tester plugin that filters the transient so the data is forced to fetch each time. 

https://github.com/joshuatf/remote-payments-tester/blob/62af802d3ab588ba22a0663958de7ac886fa0735/remote-payments-tester.php#L29

**Pull request merged**